### ISSUE TITLE
Tray Icon for Conflicts

### ISSUE THREAD

**Issue by metal450** [Martchus/syncthingtray#140]
**Relevant components**
* [x] Standalone tray application (based on Qt Widgets)
* [x] Plasmoid/applet for Plasma desktop
* [ ] Dolphin integration
* [ ] Command line tool (`syncthingctl`)
* [ ] Integrated Syncthing instance (`libsyncthing`)
* [x] Backend libraries

**Is your feature request specific to a certain platform/environment? Please specify.**
Nope

**Is your feature request related to a problem? Please describe.**
SyncThing Tray's icons are very for showing when a sync is in process, or there's some error state. However, one status that is easily missed are conflicts - I often find myself navigating into the sync dir, only to discover that there's been a conflict there for awhile that went unnoticed.  It would be very useful if the tray icon could also show an indicator when a conflict exists (reusing the "warning" indicator would be fine IMO, so as to eliminate the need for any extra UI work, though a separate/explicit icon would also be great too).  A different tray icon to indicate a conflict is really the sole feature I miss from SyncTrayzor, which is Windows only.

**Describe the solution you'd like**
When a conflict is present, show the tray icon in a warning state (or a separately configurable UI icon).

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

**Comment by Martchus** [comment]
Sounds useful, indeed. Not sure whether the official GUI shows this information (and therefore Syncthing provides an API to query it). If not, maybe it'll be better to implement it in Syncthing first (instead of doing this work on GUI level).

**Comment by metal450** [comment]
Hmm, not sure how SyncTrayzor detects it then? They actually have a whole built-in conflict resolver too, which is pretty cool (screenshot below), but perhaps a bit overkill haha.  Still, they seem to be detecting with a lot of specificity exactly what files are conflicting & where

![2022-06-04 11 26 13](https://user-images.githubusercontent.com/1877409/172020800-fcf1326a-8c6a-4a1d-9a4a-a6bb84c6dfd5.png)

↓

![2022-06-04 11 27 32](https://user-images.githubusercontent.com/1877409/172020839-1788a7bf-9d57-4b04-aa37-0bc944e71b00.png)

**Comment by Martchus** [comment]
That yellow notification doesn't look in-line with the official UI at all so I suppose it is custom code. I personally would refrain from doing any file lookups or even modifications in the GUI level. That's the job of the Syncthing backend and having two processes manipulating those files is rather messy. Doing it in the GUI also has the disadvantage that this particular feature would only work locally and if the GUI process has sufficient read/write access (which might differ from the backend process). So it makes more sense to check what the official API has to offer and perhaps create an upstream Syncthing feature request first.

**Comment by metal450** [comment]
>That yellow notification doesn't look in-line with the official UI at all

Oh yeah, that window essentially wraps the browser UI.  More or less like showing a notification at the top of a Firefox window.

> I personally would refrain from doing any file lookups or even modifications in the GUI level. That's the job of the Syncthing backend and having two processes manipulating those files is rather messy.

As far as file modifications, yeah, no need to mess with files or a resolver or anything - was just showing how that other app does it out of interest. 

> Doing it in the GUI also has the disadvantage 

Just to clarify, all I was really hoping for is a distinguishing tray icon & nothing more. Knowing that a conflict exists *somewhere* is more than sufficient (can just look at the logs or check the usual suspects to find it).  The main problem is when there's one, but I'm not aware of it.

> So it makes more sense to check what the official API has to offer and perhaps create an upstream Syncthing feature request first.

Looks like this is where SyncTrayzor's author was discussing it some years back: https://forum.syncthing.net/t/conflict-notification-via-rest-api/9255/6

"If you watch both the “conflict created” events, and the “file updated” events, you should catch all conflicts."

**Comment by Martchus** [comment]
Ok, although I'm currently not sure to which exact events (from the [documentation](https://docs.syncthing.net/dev/events.html)) these correspond to. Maybe it is also required to use the normal [REST-API](https://docs.syncthing.net/dev/rest.html) to get the initial status.

Normally I'm just checking what requests the official GUI does and do the same but it seems that the official GUI doesn't show this info so far.

**Comment by metal450** [comment]
Hmm...Looking at that, it seems the ones that would make the most sense are 

https://docs.syncthing.net/events/localchangedetected.html and 
https://docs.syncthing.net/events/remotechangedetected.html. 

Then there's a notification each time a file has changed, & it's only necessary to quickly glance at its filename to see if it contains the conflict string, right?  If it does, set the conflict icon & store the filename in hashmap (so that if a future change, whether local or remote "un-conflicts" the same file, it would remove the icon).

**Comment by metal450** [comment]
> Normally I'm just checking what requests the official GUI does and do the same but it seems that the official GUI doesn't show this info so far.

Yeah, to be honest I'm not sure how useful this would even really be in the official GUI, since most people running SyncThing just let it go in the background, & only open the UI when it's time to config.  That's why tools like SyncThing Tray & SyncTrayzor are so useful - it provides a more UI-centric way to use the software, with tray icon indicators of what's going on, without needing to bring up the whole UI :)

**Comment by Martchus** [comment]
I'm already looking for LocalChangeDetected and RemoteChangeDetected events. However, those do not help to detect conflicts that occurred before Syncthing Tray has connected. It also doesn't  look like these would be emitted for conflicts.

> Yeah, to be honest I'm not sure how useful this would even really be in the official GUI

Even if not, the backend part should still be in the backend and not in the GUI. And currently it doesn't look like the backend (Syncthing itself) provides an API for this. I'll ask on the Syncthing Matrix or IRC channel for help.

**Comment by Martchus** [comment]
The main Syncthing maintainers answer is a clear no: https://forum.syncthing.net/t/yet-another-syncthing-tray/8502/165

I suggest filing an upstream issue for such an API first.

**Comment by metal450** [comment]
Yeah, looks like SyncTrayzor shows their little icon by simply looking at the filesystem too: https://github.com/canton7/SyncTrayzor/blob/1ec7e4fadd5e7e64fb56348de630b85ccab35f07/src/SyncTrayzor/Services/Conflicts/ConflictFileManager.cs#L105

**Comment by metal450** [comment]
Fair enough. Personally, I could see an argument either way: if viewed as a frontend+backend system, you're right in that it's more logical to have all file watching take place on the backend.  However, I could also see it as unlikely to be added there, given that watching conflicts doesn't provide much use on the the "core" backend itself, & this tray app is more of a "non-core" addon.

Since you have more experience with their dev process, how are they typically with stuff like this? i.e. do requests normally get punted for years (& thus it's unlikely to see such a nice conflict tray icon available on Linux), or they're fairly responsive?

**Comment by Martchus** [comment]
I suppose it depends on the request and it's usefulness/effort ratio. Judging by Syncthing's code which I've read to search for an API I would guess the effort is quite high since Syncthing does not really keep track of these conflict files in the database after they have been created. But I could be totally wrong here.

Note that it is open source so you can already go ahead and implement it yourself (after clarifying whether a certain approach would be generally acceptable).

**Comment by metal450** [comment]
> Note that it is open source so you can already go ahead and implement it yourself

Unfortunately, far outside of my ability at this time.

> I would guess the effort is quite high

Yeah, my fear/assumption was the same. Kinda why I'd hoped there might be a willingness for something similar to SyncTrayzor, as even if not strictly ideal in terms of internal implementation, from a usability perspective the net result would be indistinguishable.  If not then not, I guess stuck just finding out about the conflicts only from the Windows side.

**Comment by metal450** [comment]
For anyone else interested in this, I cobbled together a stopgap solution to notify about conflicts on Linux, if or until this gets implemented :)

```
#!/bin/bash

monitor(){
    # Watch for files with "sync-conflict" in the name
    inotifywait --monitor --recursive -e create,moved_to --include "sync-conflict" --format '%w%f' "$1" |
    while read NEWFILE; do
        if [[ "${NEWFILE}" =~ .*tmp$ ]]; then
            : #Ignore .tmp files (every syncing file starts as tmp, so notifying for those makes every notification happen 2x)
        else
            # Use sed to parse out just the final dirname/filename
            dirAndFilename=$(echo "${NEWFILE}" | sed -E 's/(.*)\/(.*)\//\2\//')
            # Show system notification. --hint makes it appear as syncthing.
            notify-send "${dirAndFilename}" --hint='string:desktop-entry:syncthing-ui'
        fi
    done
}

monitor "/your/syncthing/folder1/" &
monitor "/your/syncthing/folder2/" &
monitor "/your/syncthing/folder3/"
```

Note: If the version of inotifywait in your distro's repo doesn't support the --include option (it was added recently), install it from source per https://unix.stackexchange.com/questions/323901/how-to-use-inotifywait-to-watch-a-directory-for-creation-of-files-of-a-specific/606213#606213

Other references:
* https://superuser.com/questions/956311/continuously-detect-new-files-with-inotify-tools-within-multiple-directories-r
* https://superuser.com/questions/1234066/replacing-only-part-of-matched-pattern-in-sed
* https://www.reddit.com/r/kde/comments/j4vm37/plasma5_is_there_a_way_to_send_a_notification/

**Comment by metal450** [comment]
no activity != solved.

**Comment by metal450** [comment]
no activity != solved.

**Comment by Martchus** [comment]
no activity -> Is someone still interested in this?

Apparently the answer is yes :-)

**Comment by cjc7373** [comment]
I think what we need here is an endpoint from syncthing that returns every file matches the pattern `*.sync-conflict-*`. Maybe an extension of `/rest/db/browse`? I'll look into that..

And then syncthingtray can monitor `LocalChangeDetected` and `RemoteChangeDetected` events to get notified of subsequent conflicts.

**Comment by Martchus** [comment]
Sounds *generally* reasonable from my side.

Of course it needed to be implemented in an efficient way. Just going though the filesystem each and every time a client connects that wants to show this kind of information is maybe not the best idea.

I suppose `LocalChangeDetected` and `RemoteChangeDetected` also needed to be extended to cover conflicts. At least I'd suspect those events are currently not emitted for conflict files. If I'm right, then I'm also not sure whether emitting those events by default for conflicts would be a good idea considering backwards compatibility. Maybe a new event type would be better.

**Comment by cjc7373** [comment]
Well I just tested that, and these events do emit for conflicted files, like:
```
[{
    "id": 1418,
    "globalID": 1665,
    "time": "2023-03-12T01:22:58.620227647+08:00",
    "type": "RemoteChangeDetected",
    "data": {
      "action": "modified",
      "folder": "xxx",
      "folderID": "xxx",
      "label": "",
      "modifiedBy": "S37NE43",
      "path": "tmp/test.sync-conflict-20230311-172258-BQ6TKV6.txt",
      "type": "file"
    }
  },
  {
    "id": 1419,
    "globalID": 1696,
    "time": "2023-03-12T01:27:46.258095946+08:00",
    "type": "LocalChangeDetected",
    "data": {
      "action": "deleted",
      "folder": "xxx",
      "folderID": "xxx",
      "label": "",
      "modifiedBy": "BQ6TKV6",
      "path": "tmp/test.sync-conflict-20230311-172258-BQ6TKV6.txt",
      "type": "file"
    }
  }]
```

**Comment by Martchus** [comment]
Good, that makes things easier (especially considering that Syncthing Tray already monitors those events in general).

**Comment by metal450** [comment]
not stale

**Comment by Kansai53** [comment]
not stale

**Comment by metal450** [comment]
not stale

**Comment by metal450** [comment]
Not sure why this auto-closed, it's not finished nor stale (per prior comment)

**Comment by murazi** [comment]
I also have the same problem. When using `SyncTrayzor`, when synchronizing note data, there will inevitably be conflicts between multiple devices. I need to manually select the files that need to be retained, but I have never seen this option in `Syncthing Tray`. Tip, I thought it had some automated optimization settings, but in the end I found that it was ignored when there were conflicting files, which caused some of my files to never be synchronized again once they conflicted. Yes, this bothers me. I think this is the most basic requirement for me to use `syncthing` and the GUI client. If this cannot be solved, there is a problem with the synchronization itself. This cannot be classified as whether the software itself is easy to use. one of the factors

**Comment by metal450** [comment]
Seems odd to have to keep commenting on this over & over & over. Shouldn't be closed unless it's actually done.

**Comment by Martchus** [comment]
It is unfortunately common that people create issues, then lose interest but never close the issue. So I added the bug to keep the tracker free of such issues.

I'm afraid nobody currently wants this feature bad enough to actually take the effort of implementing it. Maybe I'll work on it at some point but I won't promise anything.

**Comment by metal450** [comment]
> I'm afraid nobody currently wants this feature bad enough

Not sure I'd equate "nobody is coming here to proactively request it" with "nobody would benefit from it." Many product improvements (the vast majority?) don't come from direct user requests.

**Comment by Martchus** [comment]
I didn't mean to say nobody would benefit from it. I'm just observing that nobody has created a PR yet; so apparently nobody thinks this is worth the effort. And I'm on the fence myself. (Currently e.g. porting to Plasma 6 is more important to me.)

**Comment by metal450** [comment]
Sure, but if it's not an issue, it's not on the "Wish List" & would probably be forgotten?

**Comment by metal450** [comment]
not stale

**Comment by metal450** [comment]
not stale

**Comment by metal450** [comment]
not stale

**Comment by andreymal** [comment]
**@<span></span>stale** "stale" doesn't mean fixed, I guess?

**Comment by Martchus** [comment]
It just means noone is working on this or even just discussing this further. Not sure when/whether I'll implement this feature (or someone else will give it a try).

**Comment by soredake** [comment]
@Martchus why this bot even trying to close feature requests [when they are supposed to be exempt](https://github.com/Martchus/syncthingtray/blob/master/.github/stale.yml#L9-L10) from it?

**Comment by Martchus** [comment]
I don't know. I guess the exclusion worked at least at some point for "enhancement". Maybe it breaks if the tag contains a white-space. But I don't find it that bad to provoke some reactions also on feature requests to know whether someone is actually still interested in that. And if the ticket is eventually closed it doesn't really make a difference as well¹.

---

¹ Whether this is implemented or not does not depend on having an open ticket about it. It depends on someone implementing it.

**Comment by metal450** [comment]
Personally I still find it odd that I have to keep coming back & saying "not stale" over & over. I interact with a lot of projects on Github & this is the only place I've ever had to do this.

Like...there really isn't a reason a long-term wishlist needs to be constantly emptied - it's just a wishlist.

I guess maybe after a few years it could be stale or something. But needing more "reactions" every 2 months to be considered as "interest hasn't been lost" seems extremely frequent.

**Comment by Martchus** [comment]
Let's see whether using "enhancement" works.

**Comment by Martchus** [comment]
Out of curiosity I've just checked how the new conflict detection in syncthing-android works. Unfortunately it works by simply calling `find`: https://github.com/Catfriend1/syncthing-android/blob/d9d27d760ab0bc6717cd847e397a851f10ac6903/app/src/main/java/com/nutomic/syncthingandroid/util/Util.java#L525

Maybe, when Syncthing based on SQLite is released, this kind of search could be done more efficiently with SQL. Doing it on GUI wrapper level is of course still wrong.

**Comment by t214c** [comment]
@Martchus Don't you think *something* is better than nothing? I kind of understand your aesthetic aversion to hacky solutions, but the alternative is missed sync conflicts!

If you have an alternative please share.

**Comment by Martchus** [comment]
If you need this so badly, then I suggest you rather mention this in the Syncthing repo (e.g. https://github.com/syncthing/syncthing/issues/8114) because that's where this should be implemented. If none of the Syncthing core devs want to implement it, you or some other users who are interested in this feature can provide a PR. I'm not sure why I should do it considering I don't really need this feature myself.

If Syncthing itself exposes this the info (initial status via rest API + updates via event API) I can implement displaying this in Syncthing Tray (which should be rather simple then).

Note that I am usually just checking for conflicts with `find` once in a while and of course I'm trying to avoid sync conflicts in the first place.

**Comment by t214c** [comment]
Thanks. I got something working by a hacky solution on my own. I came from SyncTrayzor which had this feature so I kind of thought of it as a GUI function :)

**Comment by Martchus** [comment]
Just to make one thing clear: Syncthing Tray is not "a GUI for Syncthing". Check out the README/documentation for what Syncthing Tray is.

**Comment by andreymal** [comment]
> Syncthing Tray provides a tray icon

Tray icon is still GUI

**Comment by Martchus** [comment]
Of course it contains GUI elements and this way also partially provides a GUI for Syncthing. **However,** it is *not* an alternative/replacement for the GUI that Syncthing itself already provides - it just complements that providing quick access to the most used features. My main point is: If you want general GUI features for Syncthing it makes sense to implement them in the regular GUI of Syncthing.

**Comment by shanto** [comment]
A tray icon signaling conflicts, coupled with a customizable diff/merge command similar to what we have for browsers, could significantly enhance our workflow. @Martchus Please consider it.

This way, Syncthing Tray won't need to handle conflict resolution itself. Users will manage conflicts after file system operations, which Syncthing (the backend) will then process. Syncthing Tray remains true to its original design.
